### PR TITLE
feat: get rid of lifetime GATs

### PR DESCRIPTION
**Pull request by mattsse**
~~blocked by https://github.com/paradigmxyz/reth/pull/5476~~

* no more separate GAT traits
* no more lifetimes in StateProvider

collapsed GAT traits into the traits that inherited those

### CODE REVIEW

File: crates/storage/db/src/abstraction/database.rs [hunk 1]
```diff
@@ -1,35 +1,34 @@
...
 };
 use std::{fmt::Debug, sync::Arc};
 
-/// Implements the GAT method from:
-/// <https://sabrinajewson.org/blog/the-better-alternative-to-lifetime-gats#the-better-gats>.
+/// Trait that provides the different transaction types.
 ///
 /// Sealed trait which cannot be implemented by 3rd parties, exposed only for implementers
-pub trait DatabaseGAT<'a, __ImplicitBounds: Sealed = Bounds<&'a Self>>: Send + Sync {
+pub trait DatabaseGAT: Sealed + Send + Sync {
```
Comment (DaniPopes) [inline crates/storage/db/src/abstraction/database.rs:12]: Can rename `GAT` to `Types`

Comment (mattsse) [reply crates/storage/db/src/abstraction/database.rs:12]: +1 is also on my list

File: docs/crates/db.md [hunk 2]
```diff
@@ -268,18 +265,8 @@ This next example uses the `DbTx::cursor()` method to get a `Cursor`. The `Curso
 
 ```
 
-We are almost at the last stop in the tour of the `db` crate. In addition to the methods provided by the `DbTx` and `DbTxMut` traits, `DbTx` also inherits the `DbTxGAT` trait, while `DbTxMut` inherits `DbTxMutGAT`. These next two traits provide various associated types related to cursors as well as methods to utilize the cursor types.
-
-[File: crates/storage/db/src/abstraction/transaction.rs](https://github.com/paradigmxyz/reth/blob/main/crates/storage/db/src/abstraction/transaction.rs#L12-L17)
+We are almost at the last stop in the tour of the `db` crate. In addition to the methods provided by the `DbTx` and `DbTxMut` traits. These traits provide various associated types related to cursors as well as methods to utilize the cursor types.
```
Comment (DaniPopes) [inline docs/crates/db.md:268]: Needs rewording

File: docs/crates/db.md [hunk 3]
```diff
@@ -180,19 +168,28 @@ The `TXMut` type can be any type that implements the `DbTxMut` trait, which prov
...
+pub trait DbTxMut: Send + Sync {
+    /// Read-Write Cursor type
+    type CursorMut<T: Table>: DbCursorRW<T> + DbCursorRO<T> + Send + Sync;
+    /// Read-Write DupCursor type
+    type DupCursorMut<T: DupSort>: DbDupCursorRW<T>
+    + DbCursorRW<T>
+    + DbDupCursorRO<T>
+    + DbCursorRO<T>
+    + Send
+    + Sync;
```
Comment (DaniPopes) [inline docs/crates/db.md:180]: needs indentation

**Review by DaniPopes — CHANGES_REQUESTED** [review]
🫡 c8bc881c6aa4bd78a807e8c1c104592fcb8d33e7

File: docs/crates/db.md (hunk 3 above)
Comment (mattsse) [reply docs/crates/db.md:180]: unclear, it is formatted

Comment (DaniPopes) [reply docs/crates/db.md:180]:
```suggestion
        + DbCursorRW<T>
        + DbDupCursorRO<T>
        + DbCursorRO<T>
        + Send
        + Sync;
```

Comment (DaniPopes) [reply docs/crates/db.md:180]: Run rustfmt on this https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e1cd2ad7a9444280d2fe6ceb08158412

Comment (mattsse) [reply docs/crates/db.md:180]: ah my bad, that's a comment on the md file...

**Review by DaniPopes — APPROVED** [review]

File: crates/storage/db/src/abstraction/common.rs [hunk 4]
```diff
@@ -16,13 +18,20 @@ pub type IterPairResult<T> = Option<Result<KeyValue<T>, DatabaseError>>;
...
+    /// Sealed trait to limit the implementors of the Database trait.
     pub trait Sealed: Sized {}
-    #[allow(missing_debug_implementations)]
-    pub struct Bounds<T>(T);
-    impl<T> Sealed for Bounds<T> {}
+
+    impl<DB: Database> Sealed for &DB {}
+    impl<DB: Database> Sealed for Arc<DB> {}
+    impl Sealed for DatabaseEnv {}
+    impl Sealed for DatabaseMock {}
```
Comment (rkrasiuk) [inline crates/storage/db/src/abstraction/common.rs:32]: any good reason to keep this?

**Review by rkrasiuk — APPROVED** [review]
end of an era 🫡

File: crates/storage/db/src/abstraction/common.rs (hunk 4 above)
Comment (mattsse) [reply crates/storage/db/src/abstraction/common.rs:32]: just to retain total control, but can be removed imo

**Pull request merged**